packages/discord.js-utilities/src/lib/PaginatedMessage.ts

```diff
@@ -144,89 +152,126 @@ export class PaginatedMessage {
 		if (!this.messages.length) throw new Error('There are no messages.');
 		if (!this.actions.size) throw new Error('There are no messages.');
 
-		const firstPage = this.messages[this.index]!;
-
-		this.response ??= (await channel.send(firstPage)) as Message;
-
-		for (const id of this.actions.keys()) await this.response.react(id);
-
-		this.collector ??= this.response
-			.createReactionCollector(
-				(reaction: MessageReaction, user: User) =>
-					(this.actions.has(reaction.emoji.identifier) || this.actions.has(reaction.emoji.name)) && user.id === author.id,
-				{ idle: this.idle }
-			)
-			.on('collect', this.handleCollect.bind(this, author, channel))
-			.on('end', this.handleEnd.bind(this));
-
+		await this.setUpMessage(channel, author);
+		await this.setUpReactions(channel, author);
 		return this;
 	}
 
 	/**
-	 * This function is executed on [[PaginatedMessage.run]]. This is an extendable method.
+	 * Executed whenever [[PaginatedMessage.run]] is called.
 	 */
-	public async resolvePagesOnRun(channel: TextChannel | NewsChannel) {
+	public async resolvePagesOnRun(channel: TextChannel | NewsChannel): Promise<void> {
 		for (let i = 0; i < this.pages.length; i++) await this.resolvePage(channel, i);
 	}
 
 	/**
-	 * This function is executed whenever an action is triggered and resolved.
+	 * Executed whenever an action is triggered and resolved.
 	 * @param index The index to resolve.
 	 */
-	public async resolvePage(channel: TextChannel | NewsChannel, index: number) {
+	public async resolvePage(channel: TextChannel | NewsChannel, index: number): Promise<APIMessage> {
+		// If the message was already processed, do not load it again:
 		const message = this.messages[index];
 		if (message !== null) return message;
 
-		const page = this.pages[index];
-		const options = typeof page === 'function' ? await page(index, this.pages, this) : page;
-		const resolved = options instanceof APIMessage ? options : await new APIMessage(channel, options).resolve();
+		// Load the page and return it:
+		const resolved = await this.handlePageLoad(this.pages[index], channel, index);
 		this.messages[index] = resolved;
 
 		return resolved;
 	}
 
 	/**
-	 * This clones the current handler into a new instance.
+	 * Clones the current handler into a new instance.
 	 */
-	public clone() {
-		const clone = new PaginatedMessage({ pages: this.pages, actions: [] }).setIndex(this.index).setIdle(this.idle);
+	public clone(): PaginatedMessage {
+		const clone = new this.constructor({ pages: this.pages, actions: [] }).setIndex(this.index).setIdle(this.idle);
 		clone.actions = this.actions;
 		clone.response = this.response;
-		clone.collector = this.collector;
 		return clone;
 	}
 
-	protected async handleCollect(author: User, channel: TextChannel | NewsChannel, reaction: MessageReaction, user: User) {
+	/**
+	 * Sets up the message.
+	 * @param channel The channel the handler is running at.
+	 * @param author The author the haandler is for.
```

Comment:
q: I assume the user is there for extensions?

Reply:
Yes, it is.